diff --git a/rtl/upstream/chip/CX4/CX4.vhd b/rtl/upstream/chip/CX4/CX4.vhd
index b444ac5..8254295 100644
--- a/rtl/upstream/chip/CX4/CX4.vhd
+++ b/rtl/upstream/chip/CX4/CX4.vhd
@@ -1247,7 +1247,7 @@ begin
 		end if;
 	end process; 
 
-	DATA_ROM : entity work.spram generic map(10, 24, "rtl/chip/CX4/drom.mif")
+	DATA_ROM : entity work.spram generic map(10, 24, "rtl/upstream/chip/CX4/drom.mif")
 	port map(
 		clock		=> not CLK,
 		address	=> DATA_ROM_ADDR,
