diff --git a/rtl/upstream/main.v b/rtl/upstream/main.v
index dba1449..5b04939 100644
--- a/rtl/upstream/main.v
+++ b/rtl/upstream/main.v
@@ -1,4 +1,15 @@
-module main (
+module main #(
+	parameter reg USE_CX4 = 1'b0,
+	parameter reg USE_SDD1 = 1'b0,
+	parameter reg USE_GSU = 1'b0,
+	parameter reg USE_SA1 = 1'b0,
+	parameter reg USE_DSPn = 1'b0,
+	parameter reg USE_SPC7110 = 1'b0,
+	parameter reg USE_BSX = 1'b0,
+	parameter reg USE_MSU = 1'b0,
+	parameter reg USE_SS = 1'b0,
+	parameter reg USE_SUFAMI = 1'b0
+) (
 	input             RESET_N,
 
 	input             MCLK,
@@ -141,16 +152,6 @@ module main (
 );
 
 parameter USE_DLH = 1'b1;
-parameter USE_CX4 = 1'b1;
-parameter USE_SDD1 = 1'b1;
-parameter USE_GSU = 1'b1;
-parameter USE_SA1 = 1'b1;
-parameter USE_DSPn = 1'b1;
-parameter USE_SPC7110 = 1'b1;
-parameter USE_BSX = 1'b1;
-parameter USE_SUFAMI = 1'b1;
-parameter USE_MSU = 1'b1;
-parameter USE_SS = 1'b1;
 
 wire [23:0] CA;
 wire        CPURD_N;
